pkg/ensurance/executor/throttle.go

```diff
@@ -103,220 +63,172 @@ func (t *ThrottleExecutor) Avoid(ctx *ExecuteContext) error {
-				errPodKeys = append(errPodKeys, err.Error(), throttlePod.PodTypes.String())
-				continue
-			}
-
-			containerCPUPeriod, err := GetUsageById(throttlePod.ContainerCPUPeriods, v.ContainerId)
-			if err != nil {
-				bSucceed = false
-				errPodKeys = append(errPodKeys, err.Error(), throttlePod.PodTypes.String())
-				continue
-			}
-
-			container, err := utils.GetPodContainerByName(pod, v.ContainerName)
-			if err != nil {
-				bSucceed = false
-				errPodKeys = append(errPodKeys, err.Error(), throttlePod.PodTypes.String())
-				continue
-			}
-
-			var containerCPUQuotaNew float64
-			if utils.AlmostEqual(containerCPUQuota.Value, -1.0) || utils.AlmostEqual(containerCPUQuota.Value, 0.0) {
-				containerCPUQuotaNew = v.Value * (1.0 - float64(throttlePod.CPUThrottle.StepCPURatio)/MaxRatio)
-			} else {
-				containerCPUQuotaNew = containerCPUQuota.Value / containerCPUPeriod.Value * (1.0 - float64(throttlePod.CPUThrottle.StepCPURatio)/MaxRatio)
-			}
-
-			if requestCPU, ok := container.Resources.Requests[v1.ResourceCPU]; ok {
-				if float64(requestCPU.MilliValue())/CpuQuotaCoefficient > containerCPUQuotaNew {
-					containerCPUQuotaNew = float64(requestCPU.MilliValue()) / CpuQuotaCoefficient
+	} else {
+		ctx.ThrottoleDownGapToWaterLines, _, _ = buildGapToWaterLine(ctx.stateMap, *t, EvictExecutor{}, ctx.executeExcessPercent)
+
+		if ctx.ThrottoleDownGapToWaterLines.HasUsageMissedMetric() {
+			klog.V(6).Info("There is a metric usage missed")
+			highestPrioriyMetric := t.ThrottleDownWaterLine.GetHighestPriorityThrottleAbleMetric()
+			if highestPrioriyMetric != "" {
+				errPodKeys = t.throttlePods(ctx, &totalReleased, highestPrioriyMetric)
+			}
+		} else {
+			// The metrics in ThrottoleDownGapToWaterLines are all in WaterLineMetricsCanBeQuantified and has current usage, then throttle precisely
+			var released ReleaseResource
+			for _, m := range metricsThrottleQuantified {
+				klog.V(6).Infof("ThrottleDown precisely on metric %s", m)
+				if MetricMap[m].SortAble {
+					MetricMap[m].SortFunc(t.ThrottleDownPods)
+				} else {
+					execsort.GeneralSorter(t.ThrottleDownPods)
 				}
-			}
 
-			if limitCPU, ok := container.Resources.Limits[v1.ResourceCPU]; ok {
-				if float64(limitCPU.MilliValue())/CpuQuotaCoefficient*float64(throttlePod.CPUThrottle.MinCPURatio)/MaxRatio > containerCPUQuotaNew {
-					containerCPUQuotaNew = float64(limitCPU.MilliValue()) * float64(throttlePod.CPUThrottle.MinCPURatio) / CpuQuotaCoefficient
+				klog.V(6).Info("After sort, the sequence to throttle is ")
+				for _, pc := range t.ThrottleDownPods {
+					klog.V(6).Info(pc.PodKey.String(), pc.ContainerCPUUsages)
 				}
-			}
 
-			klog.V(6).Infof("Prior update container resources containerCPUQuotaNew %.2f, containerCPUQuota.Value %.2f,containerCPUPeriod %.2f",
-				containerCPUQuotaNew, containerCPUQuota.Value, containerCPUPeriod.Value)
+				index := 0
+				for !ctx.ThrottoleDownGapToWaterLines.TargetGapsRemoved(m) && index < len(t.ThrottleDownPods) {
```

Comment:
for index := 0; !ctx.ThrottoleDownGapToWaterLines.TargetGapsRemoved(m) && index < len(t.ThrottleDownPods); index ++ {
}